exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillConnectionImpl.java

```diff
@@ -132,10 +132,12 @@ protected DrillConnectionImpl(DriverImpl driver, AvaticaFactory factory,
             bit = new Drillbit(dConfig, serviceSet);
             bit.run();
           } catch (final UserException e) {
+            cleanup();
             throw new SQLException(
                 "Failure in starting embedded Drillbit: " + e.getMessage(),
                 e);
           } catch (Exception e) {
+            cleanup();
```

Comment:
I suppose you can use finally block to perform clean up instead of repeating `cleanup` method call after each cached exception.

Reply:
We don't want clean up to run on successful connection though, we just want to cover all the exceptions. On a successful connection, the "Curator" threads will be created but the number of threads will be proportional to number of connections, which can be be controlled by using a connection pool, hence it won't cause this issue. 

On the other hand,  calling cleanup on a successful connection may terminate that connection.  Thoughts?

Reply:
What if you use flag `doCleanup` which will be true by default and after `this.client.connect(connect, info);` succeedes, it would be set to false. In finally block you'll do clean up only if flag is true?

Reply:
One trick is this: used nested exceptions. The inner set "parses" the exception types, the outer does cleanup:

```
try {
  try {
    // do something
  } catch (ExceptionA e) {
    // Do something
  } catch (ExceptionB e) {
    // Do something else
  ... }
 } catch (Throwable t) {
    cleanup();
    throw t;
 }
```